homeassistant/components/here_travel_time/sensor.py

```diff
@@ -0,0 +1,285 @@
+        import herepy
+
+        self._hass = hass
+        self._name = name
+        self._travel_mode = travel_mode
+        self._traffic_mode = traffic_mode
+        self._route_mode = route_mode
+        self._unit_of_measurement = 'min'
+        self._response = None
+        self.valid_api_connection = True
+
+        # Check if location is a trackable entity
+        if origin.split('.', 1)[0] in TRACKABLE_DOMAINS:
+            self._origin_entity_id = origin
+        else:
+            self._origin = origin
+
+        if destination.split('.', 1)[0] in TRACKABLE_DOMAINS:
+            self._destination_entity_id = destination
+        else:
+            self._destination = destination
+
+        self._client = herepy.RoutingApi(app_id, app_code)
+        self.update()
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        if self._response is None:
+            return None
+
+        # pylint: disable=E1101
+        _summary = self._response.response['route'][0]['summary']
+        return round(_summary['trafficTime'] / 60)
+
+    @property
+    def name(self):
+        """Get the name of the sensor."""
+        return self._name
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        if self._response is None:
+            return None
+
+        # pylint: disable=E1101
+        _summary = self._response.response['route'][0]['summary']
+        # pylint: disable=E1101
+        _route = self._response.response['route']
+
+        res = {}
+        res['distance'] = _summary['distance']
+        res['distance_unit'] = LENGTH_METERS
+        res['trafficTime'] = _summary['trafficTime']
+        res['baseTime'] = _summary['baseTime']
+        res['travelTime'] = _summary['travelTime']
+        res['origin_name'] = _route[0]['waypoint'][0]['mappedRoadName']
+        res['destination_name'] = _route[0]['waypoint'][1]['mappedRoadName']
+        res[CONF_MODE] = self._travel_mode
+        res[CONF_TRAFFIC_MODE] = self._traffic_mode
+        return res
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit this state is expressed in."""
+        return self._unit_of_measurement
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    def update(self):
+        """Get the latest data from HERE."""
+        import herepy
+        # Convert device_trackers to HERE friendly location
+        if hasattr(self, '_origin_entity_id'):
+            self._origin = self._get_location_from_entity(
+                self._origin_entity_id
+            )
+
+        if hasattr(self, '_destination_entity_id'):
+            self._destination = self._get_location_from_entity(
+                self._destination_entity_id
+            )
+
+        self._destination = self._resolve_zone(self._destination)
+        self._origin = self._resolve_zone(self._origin)
+
+        # Convert location to HERE friendly location if not already so
```

Comment:
I see alot of this origin/destination stuff is duplicated from google_travel_time. this might be a good time to create helper functions in helpers folder for travel_time integrations.  This can be the start of https://github.com/home-assistant/architecture/issues/166